Single C++ file transcoder generator (and tiny Emscripten example)

cwoffenden opened the pull request:
This amalgamates the transcoder sources to a single (C++) file for quick inclusion in a project, plus adds a standalone example of using this to build for Emscripten. It's _not_ a single file header but:

- It doesn't affect the original source, it's an optional post-build step.
- It works when building as a standalone C++ file or including directly.
- The example has build options built-in to optimise output for Emscripten and iOS.
- It lowers the barrier of entry even further for adding Basis Universal to projects (one file).
- Once created it doesn't need a build system for integration.

Here's the [prebuilt Emscripten demo](https://wip.numfum.com/cw/2019-09-25/index.html). It's 269kB (of which the transcoder and texture loader support code is 249kB, and that's building with everything apart from BC7 and ATC). It's tested with various desktop, iOS, Android and ChromeOS devices.

It's based on work I did to quickly integrate [Zstandard into projects with a single file](https://github.com/facebook/zstd/tree/dev/contrib/single_file_decoder). 

(This also includes #65. The amalgamated source isn't committed, it's left to developer to build from the latest transcoder source.)

richgel999 commented:
Okay - this is very interesting! I like how it's a post-build step. I'll check this out tomorrow.
Thanks!

cwoffenden commented:
I added an option to exclude files when amalgamating, changing the example to exclude the BC7 mode 6 tables (nearly halving the final C++ output).

The changes to the transcoder source were minor and only required to silence warnings (most of the silencing going into the file used to generate the final output).

Running `combine.sh` takes a while (approx. a minute) due to making the regexes compatible with a many shells as possible. If I write them as Bash-only it's 5-6x faster. Running on Windows using the Bash bundled with Git for Windows (and I'm guessing others) takes an absolute age (but it runs!). The script was tested on Mac, Linux, BSD and Windows (and it's being used in Zstd on various Github CI servers).

I'll add something more than an Emscripten sample at some point.

cwoffenden commented:
I updated code to reflect the changes to master (mostly the enum classes in the example, and other minor tweaks). The git merge made a real mess of seeing what changed, since that's what git appears to do. I don't mind deleting this branch and recreating it, since the minor transcoder changes are in #65, and everything else is standalone. Feel free to ask.

cwoffenden commented:
I made the `-x` option in the combiner script write an `#error` if the excluded file is used, then surrounded the `basisu_transcoder_tables_bc7_m6.inc` include with guards (since it was getting included even when BC7 wasn't being built).

The only addition I think I would make is a standalone test to build the resulting file, run it to transcode, then error if it doesn't match a known output. I'd add this to the `create_translib.sh` script (since it'll add 10s to what's already a long process) to verify the amalgamated C++ builds is usable out of the box.

cwoffenden commented on transcoder/basisu_transcoder.cpp:
```diff
@@ -146,7 +146,9 @@ namespace basisu
 
 namespace basist
 {
+#if BASISD_SUPPORT_BC7_MODE6_OPAQUE_ONLY
```
Needed to stop this large file always being included.

cwoffenden commented:
I updated the script to support generating source with the same name, so now the generated files are `basis_transcoder.cpp/h`. I also added an example to the readme of generating separate amalgamated header and source (both single files).

The amalgamated header combines the existing three headers into one, with the source combining everything else except the header. The nice thing here is by keeping the `basis_transcoder` name it's a drop-in replacement for the original multi-file implementation. Why I didn't think of this before puzzles me!

richgel999 commented:
Got it - thanks a lot!
